### ISSUE TITLE
Promote GeometryTileFeature to public API

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-native#11430]
We want to expose `GeometryTileFeature` as a public API, so that filters and expressions can be evaluated by client modules.

The best way to do that would be to rename it `AbstractFeature`, convert it to vend `Geometry<int16_t>` (i.e. `mapbox::geometry::geometry<int16_t>`) as the geometry type (rather than the pre-`geometry.hpp` `GeometryCollection` type), and move it to `mbgl/util/geometry.hpp`.

cc @anandthakker @GretaCB

**Comment by anandthakker** [comment]
> convert it to vend Geometry<int16_t> (i.e. mapbox::geometry::geometry<int16_t>) as the geometry type (rather than the pre-geometry.hpp GeometryCollection type)

Capturing from chat / making sure I understand the proposal here: the idea would be that instead of decoding each feature geometry, regardless of geometry type, into `vector<vector<point>>` (aka the current `GeometryCollection`), we'd decode directly into the appropriate `mapbox::geometry` type (point/line_string/polygon/multi_*)?

**Comment by jfirebaugh** [comment]
Correct.

**Comment by anandthakker** [comment]
@springmeyer points out that this might be a good opportunity to port to https://github.com/mapbox/vtzero.

I'll plan to investigate that route first -- @kkaefer @jfirebaugh are you aware of any known hurdles?

**Comment by kkaefer** [comment]
Not aware of any hurdles, but I haven't spend a lot of time thinking this through either.

**Comment by mb12** [comment]
@kkaefer and @anandthakker As part of this, is it also possible to provide clarification regarding the difference in internal representation for the following?

-  MultiPolygon
-  Polygon with holes

GL Native returns incorrect results for queryRenderedFeatures (GL JS works correctly for the same data set). I was able to find a workaround that worked for my specific use case. It would be really nice to get a clarification on how these are represented and handle it correctly for visibleFeaturesInRect/queryRenderedFeatures .

I've included below references to the relevant issues.

https://github.com/mapbox/mapbox-gl-native/issues/10086
https://github.com/mapbox/mapbox-gl-native/issues/9686
